components/autocomplete/index.js

```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			endIndex = endTextNode ? endTextNode.nodeValue.length : 0;
 		}
-
-		if ( isOpen ) {
-			if ( match ) {
-				// Reset selection to initial when search changes
-				this.setState( {
-					selectedIndex: 0,
-					search,
-				} );
+		if ( endTextNode === null ) {
+			return null;
+		}
+		// filter the completers to those that could handle this node
+		completers = completers.filter(
+			( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+		// exit early if nothing can handle it
+		if ( completers.length === 0 ) {
+			return null;
+		}
+		let startTextNode = endTextNode;
+		let text = endTextNode.nodeValue.substring( 0, endIndex );
+		let pos = lastIndexOfSpace( text );
+		while ( pos === -1 ) {
+			const prev = onlyTextNode( startTextNode.previousSibling );
+			if ( prev ) {
+				// filter the completers to those that could handle this node
+				completers = completers.filter(
+					( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+				// exit early if nothing can handle it
+				if ( completers.length === 0 ) {
+					return null;
+				}
+				startTextNode = prev;
+				text = prev.nodeValue + text;
+				pos = lastIndexOfSpace( prev.nodeValue );
 			} else {
-				// This may occur when pressing space, for example
-				this.reset();
+				break;
 			}
-		} else if ( match ) {
-			this.setState( {
-				isOpen: true,
-				search,
-			} );
+		}
+		// exit early if nothing can handle it
+		if ( text.length <= pos + 1 ) {
+			return null;
+		}
+		// find a completer that matches
+		const open = find( completers, ( { triggerPrefix = '', allowContext = allowAnything } ) => {
+			if ( text.substr( pos + 1, triggerPrefix.length ) !== triggerPrefix ) {
+				return false;
+			}
+			const range = this.createRange( startTextNode, pos + 1, endTextNode, endIndex );
+			const before = this.createRange( container, 0, startTextNode, pos + 1 );
+			const after = this.createRange( endTextNode, endIndex, container, container.childNodes.length );
+			if ( ! allowContext( before, range, after ) ) {
+				return false;
+			}
+			return true;
+		} );
+		if ( open ) {
+			const { triggerPrefix = '' } = open;
+			const range = this.createRange( startTextNode, pos + 1, endTextNode, endIndex );
+			const query = text.substr( pos + 1 + triggerPrefix.length );
+			return { open, range, query };
+		}
+		return null;
+	}
+
+	search( event ) {
+		const { open: wasOpen } = this.state;
+		const { completers } = this.props;
+		// ensure that the cursor location is unambiguous
+		const container = event.target;
+		const cursor = this.getCursor( container );
+		// look for the trigger prefix and search query just before the cursor location
+		const match = this.findMatch( container, cursor, completers, wasOpen );
+		const { open, query, range } = match || {};
+		// create a regular expression to filter the options
+		const search = open ? new RegExp( escapeStringRegexp( query ), 'i' ) : /./;
```

Comment:
Minor and unrelated to these changes: I discovered that Lodash has a utility for this built-in, so we could probably drop the extra dependency:

https://lodash.com/docs/4.17.4#escapeRegExp

Reply:
Ok I will switch over to the lodash version.

Reply:
This is the only place we were using `escape-string-regexp` so we should remove the dependency as well.